pkg/networkserver/downlink.go

```diff
@@ -587,29 +588,35 @@ func downlinkPathsFromMetadata(mds ...*ttnpb.RxMetadata) []downlinkPath {
 		return mds[i].SNR > mds[j].SNR
 	})
 	head := make([]downlinkPath, 0, len(mds))
+	body := make([]downlinkPath, 0, len(mds))
 	tail := make([]downlinkPath, 0, len(mds))
 	for _, md := range mds {
 		if len(md.UplinkToken) == 0 || md.DownlinkPathConstraint == ttnpb.DOWNLINK_PATH_CONSTRAINT_NEVER {
 			continue
 		}
-
 		path := downlinkPath{
-			GatewayIdentifiers: md.GatewayIdentifiers,
 			DownlinkPath: &ttnpb.DownlinkPath{
 				Path: &ttnpb.DownlinkPath_UplinkToken{
 					UplinkToken: md.UplinkToken,
 				},
 			},
 		}
-		switch md.DownlinkPathConstraint {
-		case ttnpb.DOWNLINK_PATH_CONSTRAINT_NONE:
-			head = append(head, path)
-
-		case ttnpb.DOWNLINK_PATH_CONSTRAINT_PREFER_OTHER:
+		switch s := md.Source.(type) {
+		case *ttnpb.RxMetadata_GatewayIDs:
+			path.GatewayIdentifiers = s.GatewayIDs
+			switch md.DownlinkPathConstraint {
+			case ttnpb.DOWNLINK_PATH_CONSTRAINT_NONE:
+				head = append(head, path)
+			case ttnpb.DOWNLINK_PATH_CONSTRAINT_PREFER_OTHER:
+				body = append(body, path)
+			}
+		case *ttnpb.RxMetadata_PacketBroker:
 			tail = append(tail, path)
 		}
 	}
-	return append(head, tail...)
+	res := append(head, body...)
+	res = append(res, tail...)
+	return res
```

Comment:
Since we're already allocating capacities for `head`, `body` and `tail`, seems a waste not to do this:
```suggestion
	res := make([]downlinkPath, 0, len(head) + len(body) + len(tail))
	res = append(res, head...)
	res = append(res, body...)
	res = append(res, tail...)
	return res
```
clearer as well IMO

Reply:
So that does require an additional allocation, while everything is now fit in the `head` memory space